osu.Game.Rulesets.Taiko/UI/TaikoPlayfield.cs

```diff
@@ -212,12 +221,29 @@ public override void Add(DrawableHitObject h)
                     barlineContainer.Add(barline.CreateProxy());
                     break;
 
+                case DrawableDrumRoll drumRoll:
+                    drumRoll.OnHit += onDrumrollArbitraryHit;
+                    break;
+
                 case DrawableTaikoHitObject taikoObject:
                     topLevelHitContainer.Add(taikoObject.CreateProxiedContent());
                     break;
             }
         }
 
+        private void onDrumrollArbitraryHit(TaikoAction action, bool isStrong)
+        {
+            DrawableHit drawableHit;
+
+            if (action == TaikoAction.LeftRim || action == TaikoAction.RightRim)
+                drawableHit = new DrawableFlyingRimHit(Time.Current, isStrong);
+            else
+                drawableHit = new DrawableFlyingCentreHit(Time.Current, isStrong);
+
+            drumRollHitContainer.Add(drawableHit);
+            topLevelHitContainer.Add(drawableHit.CreateProxiedContent());
```

Comment:
These need to be removed on rewind. The easiest way to do this is to check `Time.Elapsed < 0` inside `Update()` and remove from these containers.

You'll need to expose a `Remove()` method from `ProxyContainer` for the proxy.
You'll also need to expose the proxied content from `DrawableTaikoHitObject` OR transform `CreateProxiedContent()` into a `GetProxiedContent()` that returns the existing proxy if there is any.

Reply:
As for the removal heuristic, since you're already giving `Time.Current` as the hitobject start time above, you could just remove all that have `StartTime > Time.Current` (i.e. if we've rewound beyond the time when the hit occurred).

Reply:
Actually on thinking about this a bit more, you can just proxy the `drumRollHitContainer` itself alongside `topLevelHitContainer` directly in the constructor, and remove the whole addition+removal of proxied content for these hits.

Reply:
@smoogipoo I'm playing with that now. How would that work exactly? Would I need to instantiate `drumRollHitContainer` as a `ProxyContainer` as well then?

Reply:
`drumRollHitContainer.CreateProxy()`, and just add it into the hierarchy alongside `topLevelHitContainer`, either above or below it.